R/sampler-adaptive.R

```diff
@@ -0,0 +1,308 @@
+    internal$scaling_history <- c(internal$scaling_history, internal$scaling)
+    internal$autocorrelation <- update_autocorrelation(
+      state$pars, internal$weight, internal$autocorrelation, pars_remove)
+    internal$mean <- update_mean(state$pars, internal$weight, internal$mean,
+                                 pars_remove)
+    internal$vcv <- update_vcv(internal$mean, internal$autocorrelation,
+                               internal$weight)
+
+    state
+  }
+
+  finalise <- function(state, model, rng) {
+    out <- as.list(internal)
+    out[c("autocorrelation", "mean", "vcv", "weight", "included",
+          "scaling_history", "scaling_weight", "scaling_increment")]
+  }
+
+  mcstate_sampler("Adaptive Metropolis-Hastings",
+                  initialise,
+                  step,
+                  finalise)
+}
+
+
+calc_scaling_increment <- function(n_pars, acceptance_target,
+                                   log_scaling_update) {
+  if (log_scaling_update) {
+    A <- -qnorm(acceptance_target / 2)
+    
+    scaling_increment <- 
+      (1 - 1 / n_pars) * (sqrt(2 * pi) * exp(A^2 / 2)) / (2 * A) + 
+      1 / (n_pars * acceptance_target * (1 - acceptance_target))
+  } else {
+    scaling_increment <- 1 / 100
+  }
+  
+  scaling_increment
+}
+
+
+qp <- function(x) {
+  outer(x, x)
+}
+
+
+calc_proposal_vcv <- function(scaling, vcv, weight, initial_vcv,
+                              initial_vcv_weight) {
+  n_pars <- nrow(vcv)
+  
+  weighted_vcv <-
+    ((weight - 1) * vcv + (initial_vcv_weight + n_pars + 1) * initial_vcv) /
+    (weight + initial_vcv_weight + n_pars + 1)
+  
+  2.38^2 / n_pars * scaling^2 * weighted_vcv
+}
+
+
+check_replacement <- function(iteration, forget_rate, forget_end) {
+  is_forget_step <- floor(forget_rate * iteration) >
+    floor(forget_rate * (iteration - 1))
+  is_before_forget_end <- iteration <= forget_end
+  
+  is_forget_step & is_before_forget_end
+}
+
+
+update_scaling <- function(scaling, scaling_weight, accept_prob,
+                           scaling_increment, min_scaling,
+                           acceptance_target, log_scaling_update) {
+  scaling_change <- scaling_increment * (accept_prob - acceptance_target) /
+    sqrt(scaling_weight)
+  
+  if (log_scaling_update) {
+    max(min_scaling, scaling * exp(scaling_change))
+  } else {
+    max(min_scaling, scaling + scaling_change)
+  }
+  
+}
+
+
+update_autocorrelation <- function(pars, weight, autocorrelation, pars_remove) {
+  if (!is.null(pars_remove)) {
+    if (weight > 2) {
+      autocorrelation <-
+        autocorrelation + 1 / (weight - 1) * (qp(pars) - qp(pars_remove))
+    } else {
+      autocorrelation <- autocorrelation + qp(pars) - qp(pars_remove)
+    }
+  } else {
+    if (weight > 2) {
+      autocorrelation <-
+        (1 - 1 / (weight - 1)) * autocorrelation + 1 / (weight - 1) * qp(pars)
+    } else {
+      autocorrelation <- autocorrelation + qp(pars)
+    }
+  }
+  
+  autocorrelation
+}
```

Comment:
this function is a little confusing for me, perhaps either a quick chat about it or some comments here, are you doing this `weight > 2` and `weight - 1` everywhere because you've already updated your `weight` before the function call? Also there is more re-scaling here, im assuming to be consistent with the unweighted vcv but perhaps would be nice to just include the rearranged forms of the $\mathbf{\Sigma}_{n}$ that you are using or just using the weighted versions and getting the unweighted version in the finalise function